java-squid/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se.symbolicvalues;
+
+import org.sonar.java.se.ConstraintManager.BooleanConstraint;
+import org.sonar.java.se.ProgramState;
+
+import javax.annotation.CheckForNull;
+
+import java.util.List;
+
+public class RelationalSymbolicValue extends BinarySymbolicValue {
+
+  public static enum Kind {
+    EQUAL("=="), NOT_EQUAL("!="), GREATER_THAN(">="), GREATER_THAN_OR_EQUAL(">="), LESS_THAN("<"), LESS_THAN_OR_EQUAL("<="), METHOD_EQUALS(".EQ."), NOT_METHOD_EQUALS(".NE.");
+    final String operand;
+
+    Kind(String operand) {
+      this.operand = operand;
+    }
+  }
+
+  private final Kind kind;
+
+  public RelationalSymbolicValue(int id, Kind kind) {
+    super(id);
+    this.kind = kind;
+  }
+
+  @Override
+  public BooleanConstraint shouldNotInverse() {
+    switch (kind) {
+      case EQUAL:
+      case METHOD_EQUALS:
+        return BooleanConstraint.TRUE;
+      default:
+        return BooleanConstraint.FALSE;
+    }
+  }
+
+  @Override
+  protected List<ProgramState> copyConstraint(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
+    ProgramState newState = programState;
+    if (programState.canReach(from) || programState.canReach(to)) {
+      newState = programState.addConstraint(this, booleanConstraint);
+    }
+    return super.copyConstraint(from, to, newState, booleanConstraint);
+  }
+
+  @Override
+  @CheckForNull
+  protected ProgramState checkRelation(BooleanConstraint booleanConstraint, ProgramState programState) {
+    RelationState relationState = binaryRelation().resolveState(programState.getKnownRelations());
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'binaryRelation' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)